k2/csrc/fsa.h

```diff
@@ -205,8 +205,37 @@ std::ostream &operator<<(std::ostream &os, const DenseFsaVec &dfsavec);
 */
 Fsa FsaFromTensor(Tensor &t, bool *error);
 
+/*
+  Create an FSA from a Array1.
+  This code will print an error message and output 'true' to 'error', and return
+  an empty FSA (with no states or arcs) if 'arc' was not interpretable as a valid FSA.
+  These requirements for a valid FSA are:
+
+    - src_state values on the arcs must be non-decreasing
+    - all arcs with -1 as the label must be to a single state (call this
+      final_state) which has no arcs leaving it
+    - final_state, if it exists, must be numerically greater than any state
+      which has arcs leaving it, and >= any state that has arcs entering it.
 
-Fsa FsaFromArray1(Array1<Arc> &arc, bool *error);
+  If there are no arcs with -1 on the label, here is how we determine the final
+  state:
+     - If there were no states at all in the FSA we'll return the empty FSA (with
+       no arcs).
+     - Else if there were some states but no arcs, we'll return the FSA (with
+       'num_states' and no arcs).
+     - Otherwise, we'll let `final_state` be the highest-numbered state
+       that has any arcs leaving or entering it, plus one.  (This FSA
+       has no successful paths but still has states.)
+
+    @param [in] arc         Source tensor.  Caution: the returned FSA will share
+                            memory with this tensor, so don't modify it afterward!
+    @param [in] num_states  The number of states for the returned FSA.
```

Comment:
I think we need to be clear about what this means... remember that there can be states which have no arcs leaving them, in the middle of other states.  Perhaps what we really mean is that the final_state is the num_states - 1?